File: tests/Varve.Benchmarks/Differential.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading;
using Varve.Conformance.Tests;
using Varve.Rdf;
using Varve.Sparql.Evaluation;

namespace Varve.Benchmarks;

/// <summary>
/// The differential run against Oxigraph (ADR 0038): every evaluation case
/// that is not blocked and does not call a SERVICE, answered by pyoxigraph
/// (<c>oxigraph/differential.py</c>) and compared with Varve's answer by the
/// conformance comparator. Varve passes every one of these cases, so a
/// disagreement is Oxigraph departing from the suite's expected result, or a
/// comparison the suite leaves open; each is reported.
/// </summary>
internal static class Differential
{
    /// <summary>Writes the cases as JSON for the Python side: query, base, and the files of each graph.</summary>
    internal static void Export(string path)
    {
        List<Dictionary<string, object?>> cases = [];
        foreach (EvaluationEntry entry in Cases())
        {
            List<string[]> graphs = [];
            graphs.AddRange(entry.Data.Select(iri => new[] { "", EvaluationSuite.PathOf(iri) }));
            HashSet<string> named = new(StringComparer.Ordinal);
            foreach (string iri in entry.GraphData)
            {
                if (named.Add(iri))
                {
                    graphs.Add([iri, EvaluationSuite.PathOf(iri)]);
                }
            }

            // As the harness does: a FROM or FROM NAMED that names a suite file loads it as that named graph.
            if (EvaluationRunner.ParseQuery(entry).Dataset is { } dataset)
            {
                foreach (RdfTerm graph in dataset.DefaultGraphs.ToArray().Concat(dataset.NamedGraphs.ToArray()))
                {
                    string iri = Encoding.UTF8.GetString(graph.Lexical);
                    if (iri.StartsWith(EvaluationSuite.PublishedRoot, StringComparison.Ordinal) && EvaluationCatalogue.Exists(iri) && named.Add(iri))
                    {
                        graphs.Add([iri, EvaluationSuite.PathOf(iri)]);
                    }
                }
            }

            cases.Add(new()
            {
                ["id"] = entry.TestIri,
                ["query"] = EvaluationSuite.PathOf(entry.QueryIri),
                ["base"] = entry.QueryIri,
                ["graphs"] = graphs,
            });
        }

        File.WriteAllText(path, JsonSerializer.Serialize(cases));
        Console.WriteLine(cases.Count + " cases written to " + path);
    }

    /// <summary>Compares Varve's answer to each case with Oxigraph's, from the directory the Python side wrote.</summary>
    internal static void Compare(string directory)
    {
        int agreed = 0;
        List<string> disagreements = [];
        List<string> refused = [];
        foreach ((EvaluationEntry entry, int index) in Cases().Select((e, i) => (e, i)))
        {
            string stem = Path.Combine(directory, index.ToString(System.Globalization.CultureInfo.InvariantCulture));
            if (File.Exists(stem + ".error"))
            {
                refused.Add(entry.TestIri + "\n    Oxigraph: " + File.ReadAllText(stem + ".error").Trim().Split('\n')[0]);
                continue;
            }

            string answer = File.Exists(stem + ".srj") ? stem + ".srj" : stem + ".nt";
            string? failure = EvaluationRunner.RunAsync(entry, EvaluationSubjects.Dataset, ValueAccess.InlineAccessor, answer, CancellationToken.None).GetAwaiter().GetResult();
            if (failure is null)
            {
                agreed++;
            }
            else
            {
                disagreements.Add(entry.TestIri + "\n    " + failure.Replace("\n", "\n    ", StringComparison.Ordinal));
            }
        }

        Console.WriteLine($"{agreed} agree, {disagreements.Count} disagree, {refused.Count} refused by Oxigraph");
        Console.WriteLine("\n## Disagreements (Varve's answer is the suite's expected one)\n");
        disagreements.ForEach(Console.WriteLine);
        Console.WriteLine("\n## Refused by Oxigraph\n");
        refused.ForEach(Console.WriteLine);
    }

    private static IEnumerable<EvaluationEntry> Cases() =>
        EvaluationCatalogue.Entries.Where(e => !EvaluationCatalogue.IsBlocked(e) && e.ServiceData.Count == 0 && !e.Suite.EndsWith("/service", StringComparison.Ordinal));
}

File: tests/Varve.Benchmarks/oxigraph/differential.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
"""Answers every exported evaluation case with pyoxigraph, for Differential.cs.

    dotnet run -c Release --project tests/Varve.Benchmarks -- --export-cases cases.json
    python differential.py cases.json answers/
    dotnet run -c Release --project tests/Varve.Benchmarks -- --differential answers/

Each case gets a fresh store; each file is one load, so blank nodes stay apart
across files as the harness keeps them. SELECT and ASK answers are written as
SPARQL JSON (N.srj), graph answers as N-Triples (N.nt), and a query Oxigraph
refuses as N.error with its message.
"""
import json
import os
import sys

import pyoxigraph as ox

FORMATS = {
    ".ttl": ox.RdfFormat.TURTLE,
    ".nt": ox.RdfFormat.N_TRIPLES,
    ".nq": ox.RdfFormat.N_QUADS,
    ".trig": ox.RdfFormat.TRIG,
    ".rdf": ox.RdfFormat.RDF_XML,
}


def answer(case, stem):
    store = ox.Store()
    for graph, path in case["graphs"]:
        fmt = FORMATS[os.path.splitext(path)[1]]
        name = ox.NamedNode(graph) if graph else ox.DefaultGraph()
        base = "https://w3c.github.io/rdf-tests/" + path.split("/tests/w3c/rdf-tests/", 1)[1]
        if fmt in (ox.RdfFormat.N_QUADS, ox.RdfFormat.TRIG):
            store.load(path=path, format=fmt, base_iri=base)
        else:
            store.load(path=path, format=fmt, base_iri=base, to_graph=name)
    with open(case["query"], encoding="utf-8") as f:
        text = f.read()
    result = store.query(text, base_iri=case["base"])
    if isinstance(result, ox.QueryTriples):
        with open(stem + ".nt", "wb") as out:
            out.write(result.serialize(format=ox.RdfFormat.N_TRIPLES))
    else:
        with open(stem + ".srj", "wb") as out:
            out.write(result.serialize(format=ox.QueryResultsFormat.JSON))


def main():
    cases = json.load(open(sys.argv[1], encoding="utf-8"))
    out = sys.argv[2]
    os.makedirs(out, exist_ok=True)
    refused = 0
    for index, case in enumerate(cases):
        stem = os.path.join(out, str(index))
        for ext in (".srj", ".nt", ".error"):
            if os.path.exists(stem + ext):
                os.remove(stem + ext)
        try:
            answer(case, stem)
        except Exception as error:  # a refusal is a result, recorded and reported
            refused += 1
            with open(stem + ".error", "w", encoding="utf-8") as f:
                f.write(type(error).__name__ + ": " + str(error))
    print(f"pyoxigraph {ox.__version__}: {len(cases)} cases, {refused} refused")


if __name__ == "__main__":
    main()
